### PR TITLE
[POC] Experimental flat index

### PR DESCRIPTION
**Pull request by eddyxu**
* Use to sort out index interface
* As the baseline for other indices because it has 100% recall.

### CODE REVIEW

**Comment by eddyxu** [comment]
With the following setup:
* CPU: AMD-5900X
* SSD: Samsung 970 Evo Plus 2TB
* 8 tokio threads
* Data:
*   1024 float32 vector, 
*    1M records,
*    Batch size=8192

Screenshot 2023-01-11 at 12 07 53 AM

FlatIndex top-1 latency is about 6s.

File: rust/src/dataset/scanner.rs [hunk 1]
```diff
@@ -138,10 +169,29 @@ impl ScannerStream {
...
+                    let batch = reader.read_batch(batch_id as i32).await.map(|b| {
+                        if with_row_id {
+                            // Add the meta "_rowid" column
+                            let mut columns = b.columns().to_vec();
+                            columns.push(Arc::new(UInt64Array::from(
+                                (row_count..row_count + b.num_rows() as u64 + fragment_id << 32)
+                                    .collect::<Vec<u64>>(),
+                            )));
+                            row_count += b.num_rows() as u64;
+                            RecordBatch::try_new(
```
Comment (changhiskhan) [inline rust/src/dataset/scanner.rs:184]: would it make sense to extend RecordBatch with append_column? so the schema and data change happens in one place.

File: rust/src/encodings/plain.rs [hunk 2]
```diff
@@ -129,13 +130,16 @@ impl<'a> PlainDecoder<'a> {
...
+        let end = self.position + array_bytes;
+        let mut ranges: Vec<Range<usize>> = vec![];
+        while start < end {
+            ranges.push(start..min(start + 1 * 1024 * 1024, end));
+            start += 1 * 1024 * 1024;
+        }
 
-        let data = self.reader.get_range(range).await?;
-        let buf: Buffer = data.into();
+        let data = self.reader.get_ranges(&ranges).await?;
```
Comment (changhiskhan) [inline rust/src/encodings/plain.rs:141]: does the inner object reader do parallel reads?

File: rust/src/encodings/plain.rs [hunk 3]
```diff
@@ -129,13 +130,16 @@ impl<'a> PlainDecoder<'a> {
...
+        let buf: Buffer = data.concat().into();
```
Comment (changhiskhan) [inline rust/src/encodings/plain.rs:142]: how expensive is this?

File: rust/src/encodings/plain.rs [hunk 4]
```diff
@@ -129,13 +130,16 @@ impl<'a> PlainDecoder<'a> {
             DataType::Boolean => bit_util::ceil(self.length, 8),
             _ => get_primitive_byte_width(self.data_type)? * self.length,
         };
-        let range = Range {
-            start: self.position,
-            end: self.position + array_bytes,
-        };
+        let mut start = self.position;
```
Comment (changhiskhan) [inline rust/src/encodings/plain.rs:133]: should we make the other encoders do this as well?

File: rust/src/index/ann/distance.rs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    // Naive implementation.
+    // TODO: benchmark and use SIMD or BLAS
+    let scores: Float32Array = (0..to.len())
+        .map(|idx| to.value(idx))
+        .map(|left| {
+            let arr = left.as_any().downcast_ref::<Float32Array>().unwrap();
+            let mut sub = subtract(arr, from).unwrap();
+            sub = multiply(&sub, &sub).unwrap();
+            sum(&sub).unwrap().sqrt()
```
Comment (changhiskhan) [inline rust/src/index/ann/distance.rs:46]: don't need to take the sqrt

File: rust/src/index/ann/flat.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            ArrowField::new("_rowid", DataType::UInt64, false),
+            ArrowField::new("score", DataType::Float32, false),
+        ]));
+        let key_arr: &Float32Array = as_primitive_array(&params.key);
+        let all_scores = stream
+            .map(|b| async {
+                if let Ok(batch) = b {
+                    let key_arr = key_arr.clone();
+                    let value_arr = batch.column_with_name(&self.column).unwrap().clone();
+                    let scores = tokio::task::spawn_blocking(move || {
```
Comment (changhiskhan) [inline rust/src/index/ann/flat.rs:90]: Rayon to parallelize dist computation inside the spawn_blocking block?

https://docs.rs/rayon/latest/rayon/iter/trait.IntoParallelRefIterator.html#tymethod.par_iter

File: rust/src/encodings/plain.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/encodings/plain.rs:141]: object_store crates doing each range in a async func, so with enough threads in tokio, it is parallel reads.

`object_store` implementation for the local file system is not high perf tho. We might need to re-implement the local SSD access later, maybe or not under the same interface. For one, it opens and closes the file descriptor for EVERY READ. quite expensive overhead, and destroys page cache. 

It is cheaper / faster than read a large chunk from local SSD, so is cheaper than S3 (which is slower than SSD).

We might want to smart about the chunking tho. make sure it does not have too many small chunks. Maybe I should limit a upper bound of # of chunks, i.e., 32?

File: rust/src/encodings/plain.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/encodings/plain.rs:142]: Relative to the 4k random access, this is not that much. 

As mentioned above, we can make this cheaper via a native local filesystem implementation (`object_store` crate is not efficient in this regard).

File: rust/src/encodings/plain.rs (hunk 4 above)
Comment (eddyxu) [reply rust/src/encodings/plain.rs:133]: Yes, certainly.

File: rust/src/index/ann/flat.rs (hunk 6 above)
Comment (eddyxu) [reply rust/src/index/ann/flat.rs:90]: It should be fine? Given this computation is done on each row in one async function. There is enough parallelism on the record level.

File: rust/src/dataset/scanner.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/dataset/scanner.rs:184]: Sure thing. lemme file a new separate / small PR just for this change.

**Comment by eddyxu** [comment]
The code has been checked in via #452 #451 #448

**Pull request closed without merging**